cypress/integration/x-msc2716-historical-import/historical-import.spec.ts

```diff
@@ -0,0 +1,288 @@
+            });
+
+            ensureVirtualUsersRegistered(synapse, AS_TOKEN, virtualUserIDs);
+        });
+    });
+
+    afterEach(() => {
+        cy.stopSynapse(synapse);
+    });
+
+    it("asdf", () => {
+        // As the application service, create the room so it is the room creator
+        // and proper power_levels to send MSC2716 events. Then join the logged
+        // in user to the room.
+        let roomId: string;
+        cy.wrap(true)
+            .then(async () => {
+                const resp = await asMatrixClient.createRoom({
+                    // FIXME: I can't use Preset.PublicChat because Cypress doesn't
+                    // understand Typescript to import it
+                    preset: "public_chat" as Preset,
+                    name: "test-msc2716",
+                    room_version: "org.matrix.msc2716v3",
+                });
+                roomId = resp.room_id;
+                return roomId;
+            })
+            .as('roomId');
+
+        cy.get<string>("@roomId").then(roomId => {
+            // Join the logged in user to the room
+            cy.joinRoom(roomId);
+
+            // Then visit the room
+            cy.visit("/#/room/" + roomId);
+        });
+
+        // Send 3 live messages as the application service.
+        // Then make sure they are visible from the perspective of the logged in user.
+        cy.get<string>("@roomId")
+            .then(async (roomId) => {
+                // Send 3 messages and wait for them to be sent
+                const liveMessageEventIds = [];
+                for (let i = 0; i < 3; i++) {
+                    // Send the messages sequentially waiting for each request
+                    // to finish before we move onto the next so we don't end up
+                    // with a pile of live messages at the same depth. This
+                    // gives more of an expected order at the end.
+                    const { event_id } = await asMatrixClient.sendMessage(roomId, null, {
+                        body: `live_event${i}`,
+                        msgtype: "m.text",
+                    });
+                    liveMessageEventIds.push(event_id);
+                }
+
+                // Make this available for later chains
+                cy.wrap(liveMessageEventIds).as('liveMessageEventIds');
+
+                // Wait for the messages to show up for the logged in user
+                waitForEventIdsInClient(liveMessageEventIds);
+
+                cy.wrap(liveMessageEventIds);
+            })
+            .then(async (liveMessageEventIds) => {
+                // Make sure the right thing was yielded
+                expect(liveMessageEventIds).to.have.lengthOf(3);
+
+                // Send a batch of historical messages
+                const insertTimestamp = Date.now();
+                const { event_ids, base_insertion_event_id } = await asMatrixClient.batchSend(roomId, liveMessageEventIds[1], null, {
```

Comment:
TODO: Move `batchSend` from `matrix-js-sdk` to just a `cy.request` in this file.

Also discussed at https://github.com/matrix-org/matrix-js-sdk/commit/d267136d26dc6b8fbaf8d764e25e194a9119b04e#r74092517

Reply:
Done ✅